File: haskell/YzuCaptcha.hs
module YzuCaptcha where

import Control.Monad
import Data.Array.IO
import Data.Bits
import Data.List
import Data.Word

import Graphics.UI.Gtk.Gdk.Pixbuf

data CharInfo = CharInfo {
        charPosX :: Int,
        charPosY :: Int,
        charWidth :: Int,
        charHeight :: Int,
        charColor :: Word32
    } deriving Show

gray, brown, white, black :: Word32
gray = 0xFFC0C0C0
brown = 0xFF13458B
white = 0xFFFFFFFF
black = 0xFF000000

filledColors :: [Word32]
filledColors = [
        0xFF0000FF, 0xFF00FF00, 0xFFFF0000,
        0xFF00FFFF, 0xFFFFFF00, 0xFFFF00FF
    ]

imgWidth, imgHeight :: Int
imgWidth = 16
imgHeight = 12

binarize :: Pixbuf -> IO Pixbuf
binarize img = do
    width <- pixbufGetWidth img
    height <- pixbufGetHeight img
    newImg <- pixbufNew ColorspaceRgb True 8 width height
    [pixels, pixels'] <- mapM pixbufGetPixels [img, newImg]

    let checkColor (p, p') = isCharColor p && isCharColor p'
        comb [] = []
        comb (x:xs) = [(x, y) | y <- xs] ++ comb xs

    forM_ [(x, y) | y <- [0..height-1], x <- [0..width-1]] $ \(x, y) -> do
        let fivePos = filter (checkBounds (width, height)) [
                    (x, y), (x-1, y), (x+1, y), (x, y-1), (x, y+1)
                ]
        (px:others) <- mapM (readArray pixels . posToIx width) fivePos
        let color = if isCharColor px
                        then black
                        else if shiftR px 24 /= 0xFF
                                then if or $ map checkColor $ comb others
                                        then black else white
                                else white
        writeArray pixels' (posToIx width (x, y)) color
    return newImg

isCharColor :: Word32 -> Bool
isCharColor px = px /= gray && px /= brown && px /= white &&
                 alpha == 0xFF
    where alpha = shiftR px 24

checkBounds :: (Int, Int) -> (Int, Int) -> Bool
checkBounds (width, height) (x, y) =
    if x >= 0 && x < width && y >= 0 && y < height
        then True else False

posToIx :: Int -> (Int, Int) -> Int
posToIx width (x, y) = x + y * width

flood :: Pixbuf -> (Int, Int) -> Word32 -> Word32 -> IO (Int, Int, Int, Int)
flood img (x, y) target filled = do
    width <- pixbufGetWidth img
    height <- pixbufGetHeight img
    pixels <- pixbufGetPixels img

    let initBounds = (x, x, y, y)
        ix = posToIx width (x, y)

    color <- readArray pixels ix
    if checkBounds (width, height) (x, y) && color == target
        then do
            writeArray pixels ix filled
            let findBounds (l, r, t, b) (x', y') = do
                    (l', r', t', b') <- flood img (x', y') target filled
                    let l'' = if l' < l then l' else l
                    let r'' = if r' > r then r' else r
                    let t'' = if t' < t then t' else t
                    let b'' = if b' > b then b' else b
                    return (l'', r'', t'', b'')
            foldM findBounds initBounds [
                    (x-1, y), (x+1, y), (x, y-1), (x, y+1),
                    (x-1, y-1), (x+1, y-1), (x-1, y+1), (x+1, y+1)
                ]
        else return initBounds

fillColor :: Pixbuf -> Word32 -> IO [CharInfo]
fillColor img target = do
    width <- pixbufGetWidth img
    pixels <- pixbufGetPixels img

    let moveX x y colors = do
            if x < width
                then do
                    px <- readArray pixels $ posToIx width (x, y)
                    if px == target
                        then do
                            let c:cs = colors
                            (l, r, t, b) <- flood img (x, y) target c
                            (next, css) <- moveX r y cs
                            let info = CharInfo (l+1) (t+1) (r-l-1) (b-t-1) c
                            return $ (info : next, css)
                        else moveX (x+1) y colors
                else return ([], colors)
        moveY _ [] = return []
        moveY colors (y:ys) = do
            (info, cs) <- moveX 0 y colors
            moveY cs ys >>= return . (info:)

    info <- liftM concat $ moveY filledColors [7, 12, 16]
    return $ filter (\i -> charHeight i >= (imgHeight `div` 2)) info

divPos :: Int -> CharInfo -> [CharInfo]
divPos num info = unfoldr findPos (info, num)
    where avg = charWidth info `div` num
          findPos (_, 0) = Nothing
          findPos (i, n) = Just (
                i { charWidth = w' }, 
                (
                    i { charPosX = charPosX i + w',
                        charWidth = w - w' },
                    n-1
                )
            )
            where w' = if avg * 2 > w then w else avg
                  w = charWidth i

locateChars :: [CharInfo] -> [CharInfo]
locateChars info
    | len < 4 = concat $ 
                    if len == 3 || len == 1 ||
                       (not $ null w2) && w1 - head w2 > 10
                        then flip map info $ \i ->
                            if w1 == charWidth i
                                then divPos num i else [i]
                        else map (divPos 2) info
    | otherwise = info
    where len = length info
          num = 4 - len + 1
          (w1:w2) = sortBy (\a b -> compare b a) ws
          ws = map charWidth info

split :: Pixbuf -> [CharInfo] -> IO [Pixbuf]
split img info = do
    pixels <- pixbufGetPixels img :: IO (PixbufData Int Word32)
    forM info $ \i -> do
        let x = charPosX i
            y = charPosY i
            w = charWidth i
            h = charHeight i
            c = charColor i
            w' = if w > imgWidth then imgWidth else w
            h' = if h > imgHeight then imgHeight else h

        new <- pixbufNew ColorspaceRgb True 8 w' h'
        pixbufCopyArea img x y w' h' new 0 0
        pixels' <- pixbufGetPixels new

        forM_ [posToIx w' (x, y) | y <- [0..h'-1], x <- [0..w'-1]] $
            \ix -> do
                px <- readArray pixels' ix
                writeArray pixels' ix $ if px == c then black else white

        new' <- pixbufNew ColorspaceRgb True 8 imgWidth imgHeight
        pixbufFill new' 0xFF 0xFF 0xFF 0xFF
        pixbufCopyArea new
            0 0 w' h' new'
            ((imgWidth - w') `div` 2) ((imgHeight - h') `div` 2)
        return new'

segment :: Pixbuf -> IO [Pixbuf]
segment img = do
    bin <- binarize img
    info <- liftM locateChars $ fillColor bin black
    chars <- split bin info
    return chars

File: ruby/yzu-captcha.rb
require 'RMagick'

module YzuCaptcha
  GRAY, BROWN, WHITE, BLACK, *FILLED_COLORS = [
    '#C0C0C0', '#8B4513', 'white', 'black',
    '#FF0000', '#00FF00', '#0000FF',
    '#FFFF00', '#00FFFF', '#FF00FF'
  ].map do |n|
    Magick::Pixel.from_color(n)
  end

  ImgWidth, ImgHeight = 16, 12

  module_function

  def segment(img)
    bin = to_bin(img)
    info = fill_color(BLACK, bin)
    new_info = locate_chars(info)
    chars = split(new_info, bin)

    return chars
  end

  def is_char_color(px)
      #luma = (
      #  px.red.to_f * 299 / 1000 +
      #  px.green.to_f * 587 / 1000 +
      #  px.blue.to_f * 114 / 1000
      #).floor

      px != GRAY && px != BROWN && px != WHITE &&
      px.opacity == 0 #&& luma < 0xC8C8
  end

  def check_bounds(width, height, x, y)
    return x >= 0 && x < width && y >= 0 && y < height
  end

  def to_bin(img)
    bin = Magick::Image.new(img.columns, img.rows)

    (0...img.rows).each do |y|
      (0...img.columns).each do |x|
        px, *sides = [
          [x, y], [x-1, y], [x+1, y], [x, y-1], [x, y+1]
        ].select do |pos_x, pos_y|
          check_bounds(img.columns, img.rows, pos_x, pos_y)
        end.map do |pos_x, pos_y|
          img.pixel_color(pos_x, pos_y)
        end

        color = 
        if is_char_color(px)
          BLACK
        elsif px.opacity != 0
          if sides.combination(2).any? {|a, b|
               is_char_color(a) && is_char_color(b)
             }
            BLACK
          else
            WHITE
          end
        else
          WHITE
        end

        bin.pixel_color(x, y, color)
      end
    end

    return bin
  end

  def flood(x, y, target, filled, img)
    left = right = x
    top = bottom = y

    if check_bounds(img.columns, img.rows, x, y)
      if img.pixel_color(x, y) == target
        img.pixel_color(x, y, filled)

        [
          [x-1, y], [x+1, y],
          [x, y-1], [x, y+1],
          [x-1, y-1], [x+1, y-1],
          [x-1, y+1], [x+1, y+1]
        ].each do |pos_x, pos_y|
          new_left, new_right, new_top, new_bottom =
          flood(pos_x, pos_y, target, filled, img)

          if new_left < left then left = new_left end
          if new_right > right then right = new_right end
          if new_top < top then top = new_top end
          if new_bottom > bottom then bottom = new_bottom end
        end
      end
    end

    return [left, right, top, bottom]
  end

  def fill_color(target, img)
    horizons = [7, 12, 16]
    num = 0
    info = []

    horizons.each do |y|
      (0...img.columns).each do |x|
        if img.pixel_color(x, y) == target
          left, right, top, bottom =
            flood(x, y, target, FILLED_COLORS[num], img)

          info << [
            [left+1, top+1, right-left-1, bottom-top-1],
            FILLED_COLORS[num]
          ]
          x = right
          num += 1
        else
          x += 1
        end
      end
    end

    return info.select do |edge, color|
      x, y, width, height = edge
      height >= ImgHeight / 2
    end
  end

  def div_position(num, char_info)
    x, y, width, height, color = char_info.flatten
    size = width / num
    new = []

    (1...num).each do
      new << [[x, y, size, height], color]
      x += size
      width -= size
    end
    new << [[x, y, width, height], color]
  end

  def locate_chars(info)
    return info if info.size >= 4

    num = 4 - info.length + 1
    w1, w2 = info.map do |edge, color|
      x, y, width, height = edge
      width
    end.sort {|i1, i2| i2 <=> i1}.take(2)

    if info.length == 3 || info.length == 1 || w1 - w2 > 10
      info.reduce([]) do |sum, ci|
        x, y, width, height = ci.first
        if w1 == width
          sum.concat(div_position(num, ci))
        else
          sum << ci
        end
      end
    else
      info.reduce([]) do |sum, ci|
        sum.concat(div_position(2, ci))
      end
    end
  end

  def split(info, img)
    info.map do |edge, color|
      x, y, char_width, char_height = edge

      if char_width > ImgWidth then char_width = ImgWidth end
      if char_height > ImgHeight then char_height = ImgHeight end

      new_px = img.get_pixels(x, y, char_width, char_height).map do |px|
        if px == color then BLACK else WHITE end
      end

      Magick::Image.
        new(ImgWidth, ImgHeight).
        store_pixels(
          (ImgWidth-char_width)/2,
          (ImgHeight-char_height)/2,
          char_width, char_height, new_px
        )
    end
  end
end
